File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/LastMatchedReciprocalCategorizer.java
package cz.fb.manaus.reactor.categorizer;

import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.categorizer.SettledBetCategorizer;
import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.SettledBet;
import org.springframework.stereotype.Component;

import java.util.Set;

@Component
public class LastMatchedReciprocalCategorizer implements SettledBetCategorizer {


    public static final String PREFIX = "lastMatchedReciprocal_";

    @Override
    public boolean isMarketSnapshotRequired() {
        return true;
    }

    @Override
    public Set<String> getCategories(SettledBet settledBet, BetCoverage coverage) {
        var marketPrices = settledBet.getBetAction().getMarketPrices();
        var reciprocal = marketPrices.getLastMatchedReciprocal();
        if (reciprocal.isPresent()) {
            if (Price.priceEq(reciprocal.getAsDouble(), 1d)) {
                return Set.of(PREFIX + "eq1");
            } else if (reciprocal.getAsDouble() > 1d) {
                return Set.of(PREFIX + "above1");
            } else {
                return Set.of(PREFIX + "bellow1");
            }
        }
        return Set.of();
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/LastMatchedReciprocalCategorizer.kt
package cz.fb.manaus.reactor.categorizer

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.categorizer.RealizedBetCategorizer
import cz.fb.manaus.core.repository.domain.Price
import cz.fb.manaus.core.repository.domain.RealizedBet
import org.springframework.stereotype.Component

@Component
class LastMatchedReciprocalCategorizer : RealizedBetCategorizer {

    override val isMarketSnapshotRequired: Boolean = true

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val marketPrices = realizedBet.betAction.marketPrices
        val reciprocal = marketPrices.lastMatchedReciprocal
        return if (reciprocal.isPresent) {
            when {
                Price.priceEq(reciprocal.asDouble, 1.0) -> setOf(PREFIX + "eq1")
                reciprocal.asDouble > 1.0 -> setOf(PREFIX + "above1")
                else -> setOf(PREFIX + "bellow1")
            }
        } else emptySet()
    }

    companion object {
        const val PREFIX = "lastMatchedReciprocal_"
    }
}
